dual_net.py

```diff
@@ -0,0 +1,155 @@
+import tensorflow as tf
+import tensorflow.contrib.layers as layers
+
+
+# A convolutional block as described in AlphaGo Zero
+def conv_block(tensor, specs):
+    tensor = layers.convolution2d(tensor,
+                                  num_outputs=specs['num_outputs'],
+                                  kernel_size=specs['kernel_size'],
+                                  stride=specs['stride'],
+                                  activation_fn=None)
+    tensor = tf.layers.batch_normalization(tensor)
+    tensor = tf.nn.relu(tensor)
+    return tensor
+
+
+# A residual block as described in AlphaGo Zero
+def residual_block(tensor, specs):
+    input_tensor = tensor
+    tensor = conv_block(tensor, specs)
+    tensor = layers.convolution2d(tensor,
+                                  num_outputs=specs['num_outputs'],
+                                  kernel_size=specs['kernel_size'],
+                                  stride=specs['stride'],
+                                  activation_fn=None)
+    tensor = tf.layers.batch_normalization(tensor)
+    tensor += input_tensor
+    tensor = tf.nn.relu(tensor)
+    return tensor
+
+
+def build_mlp(board_placeholder,
+              scope,
+              shared_layers,
+              policy_head,
+              value_head):
+    """
+    Returns the output tensors for an mlp based on the layers in shared_layers,
+    policy_head, and value_head.
+    shared_layers is a list of dicts, each dict representing a layer.
+    - Convolutional layers:
+       - d['layer'] <- 'conv'
+       - d['num_outputs'], d['kernel_size'], d['stride'] should be ints
+       - d['activation_fn'] is a function
+    - Residual layers:
+       - d['layer'] <- 'residual'
+       - other keys same as convolutional
+    - Fully connected layers:
+       - d['layer'] <- 'fc'
+       - d['num_outputs'] is an int
+       - d['activation_fn'] is a function
+    policy_head and value_head have the same structure as above but represent
+    the layers for the policy head and value head, respectively.
+
+    returns the policy output and the value output in a tuple
+    """
+    out = board_placeholder
+    for specs in shared_layers:
```

Comment:
So this loop only works if the conv blocks come first, followed by residual layers, right? Because the residual layers execute a flattening of the tensor, so you couldn't do a convolution afterwards.

I'm pointing this out because it seems this loop isn't really robust to different inputs... I wonder if we should just hardcode some of this instead of pretending that we can pass in any dictionary we want.

Reply:
Correct, as in the paper, which is one conv layer followed by a number of residual layers.  I can hardcode that.